File: src/graph/kruskal/km53.ts
import { createInterface } from "readline";

const rl = createInterface({
  input: process.stdin,
  output: process.stdout,
});

let lineno = 0;

let vertexNum = 0;
let edgeNum = 0;
let parentIdx: number[] = [];

interface IEdge {
  left: number;
  right: number;
  val: number;
}

const edges: IEdge[] = [];

let ans = 0;

rl.on("line", (line) => {
  lineno += 1;

  if (lineno === 1) {
    [vertexNum, edgeNum] = line
      .trim()
      .split(" ")
      .map((item) => Number.parseInt(item));

    parentIdx = Array.from({ length: vertexNum + 1 }, (_, idx) => idx);
    return;
  }

  const [left, right, val] = line
    .trim()
    .split(" ")
    .map((item) => Number.parseInt(item));

  edges.push({ left, right, val });

  if (lineno === 1 + edgeNum) {
    rico();
    console.log(ans);
    rl.close();
  }
});

function union(idxA: number, idxB: number) {
  const rootA = find(idxA);
  const rootB = find(idxB);
  parentIdx[rootA] = rootB;
}

function find(idx: number): number {
  const pIdx = parentIdx[idx];
  if (pIdx === idx) {
    return idx;
  }

  const root = find(pIdx);
  parentIdx[idx] = root;
  return root;
}

function rico() {
  edges.sort((a, b) => a.val - b.val);

  for (const { left, right, val } of edges) {
    const rootL = find(left);
    const rootR = find(right);
    if (rootL != rootR) {
      ans += val;
      union(rootL, rootR);
    }
  }
}

File: src/graph/kruskal/km53.py
from typing import List


vertexNum, edgeNum = map(int, input().strip().split(" "))


class Edge:
    left: int
    right: int
    val: int

    def __init__(self, left: int, right: int, val: int) -> None:
        self.left = left
        self.right = right
        self.val = val


edges: List[Edge] = []
for _ in range(edgeNum):
    left, right, val = map(int, input().strip().split(" "))
    edges.append(Edge(left, right, val))


parentIdx: List[int] = [idx for idx in range(vertexNum + 1)]


# def resetParentIdx():
#     global parentIdx
#     parentIdx = [idx for idx in range(vertexNum + 1)]


def find(idx: int) -> int:
    pIdx = parentIdx[idx]
    if pIdx == idx:
        return idx

    root = find(pIdx)
    parentIdx[idx] = root
    return root


def union(idxA: int, idxB: int) -> None:
    rootA = find(idxA)
    rootB = find(idxB)
    parentIdx[rootA] = rootB


edges.sort(key=lambda edge: edge.val)

print(edges.__dict__)

ans = 0

for edge in edges:
    rootL = find(edge.left)
    rootR = find(edge.right)

    if rootL != rootR:
        ans += edge.val
        union(rootL, rootR)


print(ans)
